Notification usage updates (#1081)
carbon-design-system/carbon-website: src/pages/components/notification/usage.mdx

## Formatting

### Anatomy

<Row>
<Column colLg={8}>

![notification anatomy image](images/notification-anatomy.png)

</Column>
</Row>

1. **Icon:** Informs users of the type of notification at a glance.
2. **Title:** Gives users a quick overview of the notification.
3. **Message (optional):** Provides additional detail and actionable steps for the user to take.
4. **Action (optional):** Ghost button that allows users to address the notification or navigates them to a page with further details.
5. **Close button (optional):** Closes the notification.

## Content

Notifications provide limited space for content, and therefore the content must be short and concise. A user should be able to quickly scan the notification, be apprised of the situation, and know what to do next.

### Main elements

#### Title

- The title should be short and descriptive, explaining the most important piece of information.
- When possible, communicate the main message using just the title.
- Use a period only if the title is a full sentence.

#### Body content

- Be concise and avoid repeating or paraphrasing the title.

> - Content should be 1-2 short sentences.

Reviewer: ```suggestion
- Limit content to one or two short sentences.
```

- Explain how to resolve the issue by including any troubleshooting actions or next steps. You can include links within the notification body that redirect the user to next steps.

#### Action

- Keep labels concise and clearly indicate the action the user can take.

> - Limit action labels to two words or less.

Reviewer: ```suggestion
- Limit action labels to one or two words. For a list of approved action labels, see Carbon's [Content guidelines](https://www.carbondesignsystem.com/guidelines/content/action-labels). 
```

**Sizing**

The width of inline notifications varies based on content and layout. They can expand to the fill the container or content area they relate to. Their height is based on the content length, which should not exceed two lines of text.

**Placement**

Inline notifications appear near their related items. In forms, we recommend placing the inline notification at the bottom of the form, right before the submission buttons. Often error notifications are used in forms to accompany incorrect invalid text inputs. Depending on the context of the page, inline notifications can appear above the content as well.

<Row>
<Column colLg={8}>

![Example of inline notification placement](images/notification-placement-inline.png)

<Caption>
  Example of an inline notification supporting an error message on a form
</Caption>

</Column>
</Row>

### Behavior

**Dismissal**

Inline notifications do not dismiss automatically. They persist on the page until the user dismisses them or takes action that resolves the notification.

> A small "x" in the top right corner is used to dismiss inline notifications. Including the close button is optional and should not be included in a notification if it is critical a user reads or interacts with it.

Reviewer: ```suggestion
A small "x" in the top right corner is used to dismiss inline notifications. Including the close button is optional and should not be included if it is critical for a user to read or interact with the notification.
```

## Toast notifications

### Formatting

**Action**

Toast notifications can include a link at the end of their body content. This link should be short and navigate users to a page or modal where they can take action to address the notification or find further information. Because toast notifications automatically dismiss, it is important that there are alternative routes to navigate to the link destination.

If you include links within the notification body that redirect the user to steps, be sure keyboard and screenreader users have enough time to access the notification or can easily find the notification after it disappears.

> **Timestamp**

Reviewer: ```suggestion
**Time stamp**
```

> Toast notifications include a timestamp at the bottom the container. The timestamp shows the time the notification was sent. Using timestamps is optional but toasts should be consistent across the product so either all toasts should include timestamps or none of them should. The timestamp is optional and can be removed if a third line of content is needed.

Reviewer: ```suggestion
Toast notifications include a time stamp at the bottom the container. The time stamp shows the time the notification was sent. Using time stamps is optional but toasts should be consistent across the product so either all toasts should include time stamps or none of them should. The time stamp is optional and can be removed if a third line of content is needed.
```

</Column>
</Row>

### Behavior

**Dismissal**

Toasts dismiss automatically after five seconds on the screen. They can also include a close button so users can dismiss them sooner. Toasts cover content on the screen so they should always be easily dismissed. Because toast notifications dismiss automatically, users should be able to access them elsewhere after the toast disappears. This allows them to be accessible for users who need more time reading the notification or who may want to refer back to the notification.

## Modifiers

### High and low contrast

Carbon supports high and low contrast style notifications. High-contrast notifications are best for critical messaging while low-contrast notifications are less visually disruptive for users.

It's up to the product team to decide which notification style to use in their product. Inline and toast notifications can use different styles but you should never mix styles within each notification type. When in doubt, use low-contrast notifications.

<Row>
<Column colLg={8}>

![High and low contrast inline notifications](images/notification-usage-4.png)

</Column>
</Row>

## Related

> - Modals

Reviewer: ```suggestion
- [Modal component](https://www.carbondesignsystem.com/components/modal/usage/)
```

> - Notification pattern

Reviewer: ```suggestion
- [Notifications pattern](https://www.carbondesignsystem.com/patterns/notification-pattern)
```

</Row>

## Behavior

### Dismissal

We recommend that toast notifications automatically disappear after five seconds. Inline notifications are persistent until the user dismisses them. All notifications have at least one method of dismissal which is typically a small “x” in the upper right hand corner.

### Overflow

Notification messages should be kept short. For toast notifications, the timestamp is optional and can be removed if a third line of content is needed.

If a toast or inline notification requires a message of longer than three lines, include a short message with a “View more” link that takes the user to a view of the full notification message. This can be either a full page with more details or a modal.

## Content

Notifications provide limited space for content, and therefore the content must be concise. A user should be able to quickly scan the notification, be apprised of the situation, and know what to do next.

### Title

- The title should be short and descriptive, explaining the most important piece of information.
- When possible, communicate the main message using just the title.
- Use a period only if the title is a full sentence.

### Body content

> - The body of the notification should be no more than two lines.

Reviewer 2: How are we suggesting folks tackle this in the globalization/overflow use-case?

Reviewer 3: Yeah good point with translation we can't really control the line length. We could make these comments less line specific. Like "Content should be 1-2 short sentences." I do think we should emphasize somehow that it shouldn't have excess content. I could even say one short sentence to emphasize that.

- Be concise and avoid repeating or paraphrasing the title.

> - Explain how to resolve the issue by including any troubleshooting actions or next steps. You can include links within the notification body that redirect the user to next steps.

Reviewer 2: How are we anticipating keyboard or screen reader users to navigate to the toast if there is interactive content inside? By default, it seems like these might be hard to reach quickly within a small time window.

Reviewer 3: I think this is more talking to inline notification that toast. But i think it would be good to emphasis that point here. Like 

> "If you include links within the notification body that redirect the user to steps, be sure keyboard and screenreader user have enough time to access the notification or can easily find the notification after it disappears."

### Action

> - Keep labels consice and clearly indicate the action the user can take.

Reviewer 2: ```suggestion
- Keep labels concise and clearly indicate the action the user can take.
```

- Limit action labels to two words or less.

### Timestamp

- The timestamp reflects the time the toast notification was sent. Timestamps should be localized to the user's timezone or preferences.

For further content guidance, see Carbon's [content guidelines](/guidelines/content/general).

## Modifiers

### High and low contrast

Carbon supports high and low constrast style notifications. High-contrast notifications are best for critical messaginging while low-contrast notifications are less visually disruptive for users.

It's' up to the product team to decide which notification style to use in their product. Inline and toast notifications can use different styles but you should never mix styles within each notification type. When in doubt, use low-contrast notifications.

<Row>
<Column colLg={8}>

![High and low contrast inline notifications](images/notification-usage-4.png)

</Column>
</Row>

### Action

**Inline notification**

> Inline notifications have an optional ghost button action that is right aligned. This button should allow users to take further action on the notification.

Reviewer 2: > Inline notifications have an optional ghost button action that is **right aligned**

What would be the best way to talk about direction if we do left-to-right and right-to-left? Or is this is saying that the action is always on the right-hand side?

Reviewer 3: We dont really address this topic anywhere and at the moment we dont really have a strategy for it. This is something we should talk about though.

Reviewer 3: Speaking of the action button placement though this week we are adding a small width behavior change to it. So it sometime also on the bottom.
https://github.com/carbon-design-system/carbon/pull/5763

**Toast notifications**

> Toast notifications can include a link at the end of their body content. This link should be short and navigate users to a page or modal where they can take action to address the notification or find further information. Because toast notifications automatically dismiss, it is important that there are alternative routes to navigate to the link destination.

Reviewer 2: > Because toast notifications automatically dismiss, **it is important** that there are alternative routes to navigate to the link destination.

Should we potentially include stronger language around this in the sense that they must provide an alternate way to view the notification?

3. **Message (optional):** Provides additional detail and actionable steps for the user to take.
4. **Action (optional):** Ghost button that allows users to address the notification or navigates them to a page with further details.
5. **Close button (optional):** Closes the notification.

## Content

Notifications provide limited space for content, and therefore the content must be short and concise. A user should be able to quickly scan the notification, be apprised of the situation, and know what to do next.

### Main elements

#### Title

- The title should be short and descriptive, explaining the most important piece of information.
- When possible, communicate the main message using just the title.
- Use a period only if the title is a full sentence.

#### Body content

- Be concise and avoid repeating or paraphrasing the title.
- Limit content to one or two short sentences.
- Explain how to resolve the issue by including any troubleshooting actions or next steps. You can include links within the notification body that redirect the user to next steps.

#### Action

- Keep labels concise and clearly indicate the action the user can take.

> - Limit action labels to one or two words. For a list of approved action labels, see Carbon's [content guidelines](https://www.carbondesignsystem.com/guidelines/content/action-labels).

Reviewer 2: Was surprised to see approved here, is there a process through which teams get labels approved? Or are these recommendations by our team?

Reviewer: These labels have been developed through multiple rounds of review with Cloud Data & AI, IoT, and Watson designers and content designers.

Reviewer 4: Suggestion - should read; "For a list of recommended labels...". I'm not sure "approved" is correct here, in that it implies anything else is non-approved (but we don't have any real control over the action labels different teams might use).

Process for choosing which labels are recommended is outside the scope of this PR, but we can discuss if there is something we want to document/clarify.

Reviewer 3: I don't think it's implying that labels have to be approved before using, but merely saying the action labels on that content page are approved to use. Maybe it should say "For a list of _suggested_ action labels, ..."

Reviewer: It needs to be stronger than just suggested. If people have concerns with "approved" then I'd like to go with "recommended".

Reviewer 3: oops sorry y'all my internet is terrible and for some reason no one else comments loaded so I wasn't purposefully ignoring any earlier comments.

<Row>
<Column colLg={8}>

![Example of inline notification placement](images/notification-placement-inline.png)

<Caption>
  Example of an inline notification supporting an error message on a form
</Caption>

</Column>
</Row>

### Behavior

#### Dismissal

Inline notifications do not dismiss automatically. They persist on the page until the user dismisses them or takes action that resolves the notification.

A small "x" in the top right corner is used to dismiss inline notifications. Including the close button is optional and should not be included if it is critical for a user to read or interact with the notification.

## Toast notifications

### Formatting

<Row>
<Column colLg={8}>

![Example of toast notification with action button and time stamp](images/notification-toast-action.png)

</Column>
</Row>

#### Action

Toast notifications can include a link at the end of their body content. This link should be short and navigate users to a page or modal where they can take action to address the notification or find further information. Because toast notifications automatically dismiss, there should be alternative routes to navigate to the link destination.

> If you include links within the notification body that redirect the user to steps, be sure keyboard and screenreader users have enough time to access the notification or can easily find the notification after it disappears.

Reviewer 2: In general, it seems like we would want to recommend that notifications don't have any interactive content within them. When using a live region, the text of the area is what is communicated to the end user:

![demo](https://user-images.githubusercontent.com/3901764/78923406-c7272500-7a5d-11ea-9398-24083ab5dd95.gif)

As a result, there is no way for them to know if the content is interactive or not nor is there an easy way for them to navigate to it.

Reviewer 4: More specifically. Is the feedback here that this guidance should change because we don't/can't include links in notifications?

Reviewer 3: This is technically speaking something that we currently allow and show in the storybook. She is only commenting on what we're currently allowing. If we don't want to allow this then we need to also remove this from the component code. 

![image](https://user-images.githubusercontent.com/11670886/78930664-1bd09d00-7a6a-11ea-9308-cc78a6a66267.png)

Reviewer 2: @aagonzales definitely agreed, I think this needs to be a conversation for us all to better understand what is/is not applicable here. It feels odd given that so many sites use this pattern that this is how a screen reader behaves, maybe I'm misunderstanding something? Was surprised when I was testing it out in that recording above that it announced the way that it did.

@dakahn do you have any insights here? Clearly this is a common pattern, including notifications at the OS-level, is there anything we can do to accommodate interactive content either through a link/action button/close button?

Reviewer 5: **At IBM we have the requirement that any mouse interaction have a keyboard equivalent.** With notifications that popup and disappear (toasts or whatever) and have interactable child elements this rule becomes problematic. 

How do we move keyboard focus to that notification? If this were a modal dialog we'd simply steal the users focus and _move it to the dialog_ when the modal was opened and then trap focus in the dialog until an action was taken. Toasts come and go as the user works and are intended to be unobtrusive so the above described interaction pattern doesn't really work. 

At the application level there are ways to solve around this problem (having notifications collected somewhere that _is keyboard accessible_ like this:

![image](https://user-images.githubusercontent.com/40970507/78946424-4af60700-7a88-11ea-98bc-2810eb0c7435.png)

but its always highly inadvisable to use operating system level interactions as guidance for building websites since -- in short -- they are very different contexts and only superficially comparable.

Reviewer 2: @dakahn it seems like the guidance from WAI-ARIA gets distilled into two categories: alerts and alert dialogs:

> An alert is an element that displays a brief, important message in a way that attracts the user's attention without interrupting the user's task
> Because alerts are intended to provide important and potentially time-sensitive information without interfering with the user's ability to continue working, it is crucial they do not affect keyboard focus.

> The alert dialog is designed for situations where interrupting work flow is necessary.

---

Alerts are what we can use for time-based notifications, but if there is anything interactive then we need to use the alert dialog?

What happens if we want to use a timed alert, but also want to give someone the ability to dismiss it early? This seems like it'd directly contradict the alert guidance as it says that it shouldn't steal focus from the user.

**Links & resources**

- [Alert example](https://www.w3.org/TR/wai-aria-practices/examples/alert/alert.html)
- [Alert dialog example](https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/alertdialog.html)

Reviewer 6: Most common example I can think of is the service work content refresh confirmation or a cookie approval form.

A notification tray isn’t enough for these types of product-level notifications/approvals. A modal on the other hand is far too invasive. I don’t see the problem with adding actions to our Toast notifications so long as we prevent their automatic dismissal.

Reviewer 5: > What happens if we want to use a timed alert, but also want to give someone the ability to dismiss it early?

They're highly problematic for sure. You can minimize a lot of these gnarly UX frustrations for keyboard/screen reader/magnification users, but some of those solutions cause frustrations for other (dialogs with interactable elements becoming modal for example solves one problem but creates another).

#### Time stamp

Toast notifications include a time stamp at the bottom the container. The time stamp shows the time the notification was sent. Using time stamps is optional but toasts should be consistent across the product so either all toasts should include time stamps or none of them should. The time stamp is optional and can be removed if a third line of content is needed.

#### Sizing

Toast notifications use a fixed width and their height depends on the length of the notification message. As noted in the content guidelines, limit toast notifications to two lines of text.

#### Placement

> Toast notifications slide in and out from the top right of the screen. They align to the edge of the last column and can stack with `$spacing-03` in-between. New toast notifications should appear at the top of the list, with older notifications being pushed down until they are dismissed.

Reviewer 2: What happens if the height of the stacked notifications is greater than the browser height?

Notifications are typically absolutely positioned, as a result, how would one implement this guidance:

> They align to the edge of the last column

Reviewer 4: Clarification on @joshblack's last point; "absolute positioning" means that an element doesn't reference other elements, so the "align to the edge of the last column" is difficult/impossible to apply from a code perspective.

Do we need to either address the design guidance, or address the development implementation?

Reviewer 3: This is a good point that we don't currently have a solution for. This will need to be opened as a issue to figure out. I also think if they're disappearing after x seconds the chances of there being a long stack of them seems very unlikely.

Author: @joshblack do you have a suggestion for how to provide the positioning guidance? I was trying to address this issue https://github.com/carbon-design-system/carbon-website/issues/76

Reviewer 2: @jillianhowarth it seems like we could re-word it to say that the notifications are aligned vertically and use the spacer in between and not mention the edge alignment? Just a thought 👀

> Actionable toast notifications do not appear on mobile screen widths.

Reviewer 2: What was the reasoning behind this constraint? 👀

Author: This is existing guidance we have right now. I'm not sure the original reasoning behind it but I'm trying not to changing existing guidance without having a pretty good reason to change it.

Reviewer 2: Got it! Makes sense 👍

<Row>
<Column colLg={8}>

![Example of toast notification placement](images/notification-placement-toast.png)

</Column>
</Row>

### Behavior

#### Dismissal

> Toasts dismiss automatically after five seconds on the screen. They can also include a close button so users can dismiss them sooner. Toasts cover content on the screen so they should always be easily dismissed. Because toast notifications dismiss automatically, users should be able to access them elsewhere after the toast disappears. This allows them to be accessible for users who need more time reading the notification or who may want to refer back to the notification.

Reviewer 2: > Because toast notifications dismiss automatically, users should be able to access them elsewhere after the toast disappears

Should we word this more strongly to say that there must be an alternate location if the toast is automatically dismissed?

Reviewer 4: It definitely seems like it should be a requirement rather than a preference, right? (so "users _must_ be able to access..." rather than "users _should_...")

Reviewer 3: Would simply changing "should be" to "must be" be enough?

Reviewer: Yes, I think so.
